common/txmgr/address_state.go

```diff
@@ -224,23 +286,194 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUn
 }
 
 // moveTxToFatalError moves a transaction to the fatal error state.
+// It returns an error if there is no transaction with the given ID.
+// It returns an error if the transaction is not in an expected state.
+// Unknown transaction states will cause a panic this includes Unconfirmed and Confirmed transactions.
 func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveTxToFatalError(
 	txID int64, txError null.String,
 ) error {
+	as.Lock()
+	defer as.Unlock()
+
+	tx := as.allTxs[txID]
+	if tx == nil {
+		return fmt.Errorf("move_tx_to_fatal_error: no transaction with ID %d", txID)
+	}
+
+	as._moveTxToFatalError(tx, txError)
+
+	switch tx.State {
+	case TxUnstarted:
+		_ = as.unstartedTxs.RemoveTxByID(txID)
+	case TxInProgress:
+		as.inprogressTx = nil
+	case TxConfirmedMissingReceipt:
+		delete(as.confirmedMissingReceiptTxs, tx.ID)
+	default:
+		panic("move_tx_to_fatal_error: unknown transaction state")
```

Comment:
nit: could you include the state we panic on in the message for easier debugging?